File: IRNDCG.java
import java.util.*;

public class IRNDCG{
	public static void main(String [] args){
	ArrayList<Integer> relevance = new ArrayList<>(); // stops at the N value
	// add relevance values in order of ranking
	Scanner keyboard = new Scanner(System.in);
	System.out.print("Enter N value: "); int N = keyboard.nextInt(); 
	System.out.println("Enter N relevance values:");
	for(int a = 0; a < N; a++){
		int rel = keyboard.nextInt();
		relevance.add(rel);	
	}
		
	double DCG = 0; 
	int j = 1;

	// DCG value
	for(int i = 0; i < relevance.size(); i++)
	{
		DCG = DCG + (relevance.get(i)/(Math.log(j+1) / Math.log(2)));
		j = j + 1;
	}
	System.out.println("DCG value:" + DCG);

	// IDCG value...sort to ideal order.
	Collections.sort(relevance);
	Collections.reverse(relevance);
	System.out.println("Ideal order" + relevance);
	double IDCG = 0; 
	int k = 1;
	for(int i = 0; i < relevance.size(); i++)
	{
		IDCG = IDCG + (relevance.get(i)/(Math.log(k+1) / Math.log(2)));
		k = k + 1;
	}
	System.out.println("IDCG value:" + IDCG);

	// NDCG value
	double NDCG = DCG / IDCG;
	System.out.println("NDCG value:" + NDCG);
	}
}

File: irndcg.py
import math
def irndcg(query_id, relevant_doc_ids, query_file):
    path = query_file[0:query_file.rfind('\\')]
    relevant_doc_ids = [int(i) for i in relevant_doc_ids]
    relevance_file = path+"\\relevance."+query_id #to fetch relevance values from relavance file4
    relevance = []
    print("relpath:",relevance_file)
    f =  open(relevance_file, "r")
    content = f.read()
    split_content = content.splitlines()
    for i in range(len(split_content)): #searching in relevance file for the relevance of each document
        if i in relevant_doc_ids:
            relevance.append(int(split_content[i]))
            #print(i)
            #print(split_content[i])
    #print(relevance)

    dcg = 0
    j = 1
    #dcg value
    for i in range(len(relevance)):
        dcg = dcg + (relevance[i]/(math.log(j+1)/math.log(2)))
        j = j + 1
    print("DCG value: " + str(dcg))
    #idcg value  .... sort to ideal order.
    relevance.sort()
    relevance.reverse()
    print("Idea; order" + str(relevance))

    idcg = 0
    k = 1
    for i in range(len(relevance)):
        idcg = idcg + (relevance[i]/(math.log(k+1)/math.log(2)))

        k = k +1
    print("IDCG value: "+ str(idcg))

    #NDCG value
    try:
        ndcg = dcg/idcg
    except:
        ndcg = 0

    print("NDCG value: " + str(ndcg))

    return ndcg
